docs/architecture/adr-047-extend-upgrade-plan.md

```diff
@@ -0,0 +1,206 @@
+  Tools utilizing these `UpgradeInstructions` MUST fail if a `checksum` is provided but is different from the checksum of the result returned by the `url`.
+- `checksum_algo` is a string identify the algorithm used to generate the `checksum`.
+  Recommended algorithms: `sha256`, `sha512`.
+  Algorithms also supported (but not recommended): `sha1`, `md5`.
+  If a `checksum` is provided, a `checksum_algo` MUST also be provided.
+
+A `url` is not required to contain a `checksum` query parameter.
+If the `url` does contain a `checksum` query parameter, the `checksum` and `checksum_algo` fields MUST also be populated, and their values MUST match the value of the query parameter.
+For example, if the `url` is `"https://example.com?checksum=md5:d41d8cd98f00b204e9800998ecf8427e"`, then the `checksum` field must be `"d41d8cd98f00b204e9800998ecf8427e"` and the `checksum_algo` field must be `"md5"`.
+
+### Upgrade Module Updates
+
+If an upgrade `Plan` does not use the new `UpgradeInstructions` field, existing functionality will be maintained.
+
+We will update the creation of the `upgrade-info.json` file to include the `UpgradeInstructions`.
+
+We will update the optional validation available via CLI to account for the new `Plan` structure.
+We will add the following validation:
+1.  If `UpgradeInstructions` are provided:
+    1.  There MUST be at least one entry in `artifacts`.
+    1.  All of the `artifacts` MUST have a unique `platform`.
+    1.  For each `Artifact`, if the `url` contains a `checksum` query parameter:
+        1. The `checksum` query parameter value MUST be in the format of `{checksum_algo}:{checksum}`.
+        1. The `{checksum}` from the query parameter MUST equal the `checksum` provided in the `Artifact`.
+        1. The `{checksum_algo}` from the query parameter MUST equal the `checksum_algo` provided in the `Artifact`.
+1.  The following validation is currently done using the `info` field. We will apply similar validation to the `UpgradeInstructions`.
+    For each `Artifact`:
+    1.  The `platform` MUST have the format `{OS}/{CPU}` or be `"any"`.
+    1.  The `url` field MUST NOT be empty.
+    1.  The `url` field MUST be a proper URL.
+    1.  A `checksum` MUST be provided either in the `checksum` field or as a query parameter in the `url`.
+    1.  If the `checksum` field has a value and the `url` also has a `checksum` query parameter, the two values MUST be equal.
+    1.  The `url` MUST return either a file or an archive containing either `bin/{DAEMON_NAME}` or `{DAEMON_NAME}`.
+    1.  If a `checksum` is provided (in the field or as a query param), the checksum of the result of the `url` MUST equal the provided checksum.
+
+Downloading of an `Artifact` will happen the same way that URLs from `info` are currently downloaded.
+
+### Cosmovisor Updates
+
+If the `upgrade-info.json` file does not contain any `UpgradeInstructions`, existing functionality will be maintained.
```

Comment:
```suggestion
Current meaning of `Plan.Info` will be deprecated - the attribute will continue to exist, but it will have a non machine interpretable content.
    * If the `upgrade-info.json` file contains `UpgradeInstructions` then `Plan.Info` won't be used.
    *  otherwise, hypervisors are expected to fallback to the existing behavior and try to interpret `Plan.Info`.
```

Reply:
I'm going to resolve this conversation so it's not happening in two places.

Here's the other conversation: https://github.com/cosmos/cosmos-sdk/pull/10602#discussion_r782472640